apps/opencs/view/render/terrainshapemode.cpp

```diff
@@ -0,0 +1,1459 @@
+    }
+
+    if (mBrushShape == CSVWidget::BrushShape_Square)
+    {
+        for(int i = vertexCoords.first - r; i <= vertexCoords.first + r; ++i)
+        {
+            for(int j = vertexCoords.second - r; j <= vertexCoords.second + r; ++j)
+            {
+                cellId = CSMWorld::CellCoordinates::vertexGlobalToCellId(std::make_pair(i, j));
+                cellCoords = CSMWorld::CellCoordinates::fromId(cellId).first;
+                int x = CSMWorld::CellCoordinates::vertexGlobalToInCellCoords(i);
+                int y = CSMWorld::CellCoordinates::vertexGlobalToInCellCoords(j);
+                if (mShapeEditTool == ShapeEditTool_Drag) alterHeight(cellCoords, x, y, mTotalDiffY);
+                if (mShapeEditTool == ShapeEditTool_PaintToRaise || mShapeEditTool == ShapeEditTool_PaintToLower)
+                {
+                    alterHeight(cellCoords, x, y, mShapeEditToolStrength);
+                    float smoothMultiplier = static_cast<float>(CSMPrefs::get()["3D Scene Editing"]["landedit-post-smoothstrength"].toDouble());
+                    if (CSMPrefs::get()["3D Scene Editing"]["landedit-post-smoothpainting"].isTrue()) smoothHeight(cellCoords, x, y, mShapeEditToolStrength * smoothMultiplier);
+                }
+                if (mShapeEditTool == ShapeEditTool_Smooth) smoothHeight(cellCoords, x, y, mShapeEditToolStrength);
+                if (mShapeEditTool == ShapeEditTool_Flatten) flattenHeight(cellCoords, x, y, mShapeEditToolStrength, mTargetHeight);
+            }
+        }
+    }
+
+    if (mBrushShape == CSVWidget::BrushShape_Circle)
+    {
+        for(int i = vertexCoords.first - r; i <= vertexCoords.first + r; ++i)
+        {
+            for(int j = vertexCoords.second - r; j <= vertexCoords.second + r; ++j)
+            {
+                cellId = CSMWorld::CellCoordinates::vertexGlobalToCellId(std::make_pair(i, j));
+                cellCoords = CSMWorld::CellCoordinates::fromId(cellId).first;
+                int distanceX = abs(i - vertexCoords.first);
+                int distanceY = abs(j - vertexCoords.second);
+                float distance = sqrt(pow(distanceX, 2)+pow(distanceY, 2));
+                int x = CSMWorld::CellCoordinates::vertexGlobalToInCellCoords(i);
+                int y = CSMWorld::CellCoordinates::vertexGlobalToInCellCoords(j);
+                float distancePerRadius = 1.0f * distance / r;
+                float smoothedByDistance = 0.0f;
+                if (mShapeEditTool == ShapeEditTool_Drag) smoothedByDistance = mTotalDiffY - mTotalDiffY * (3 * distancePerRadius * distancePerRadius - 2 * distancePerRadius * distancePerRadius * distancePerRadius);
+                if (mShapeEditTool == ShapeEditTool_PaintToRaise || mShapeEditTool == ShapeEditTool_PaintToLower) smoothedByDistance = (r + mShapeEditToolStrength) - (r + mShapeEditToolStrength) * (3 * distancePerRadius * distancePerRadius - 2 * distancePerRadius * distancePerRadius * distancePerRadius);
```

Comment:
Not sure what this expression does but this could be a function with explaining name. Another reason is that this expression repeats at least twice in a slightly different form.

Reply:
You probably mean the smoothedByDistance here? It could be a function, but because it's only there, is it really worth it? Is it easier to figure out the function just by looking at the formula?

What it actually does: It uses a formula `y = 3x^2 - 2x^3`, where x is the distance from the center of the circle tool. This formula creates a "bump-shape", when using circular tool to change land heights. It gives positive values though, so instead of circular tool making a change for the whole area with it's full strength, the inverse of this bump shape formula is subtracted from the height changes, depending on how far from the tool center the change takes place.

For the `if`, I kind of prefer the current way, as I find it the most clear way to state what the code does. `mShapeEditTool` is the shape editing tool that's currently used by the user, values are enums of different tools.

Reply:
Yes, I mean `smoothedByDistance`. `y = 3x^2 - 2x^3` is used twice (actually `dy - dy * (3x^2 - 2x^3)`). If someone decides this formula is not good enough then two changes are required. One might be forgotten.